### PR TITLE
XRPL NFT Bridging TRN to XRPL Node changes

### PR DESCRIPTION
**Pull request by surangap**
### Changelog
 - Add `Xls20Ext` and plug to `pallet-xrpl-bridge` config to interact with Xls20 pallet.
 - Update `NFTExt` and plug to `pallet-xrpl-bridge` to interact with NFT pallet.
 - Add `withdraw_nft()` extrinsic to support NFT withdrawal from TRN to XRPL
 - Move `CrossChainCompatibility` to `seed-primitives`

### CODE REVIEW

File: pallet/xls20/src/lib.rs [hunk 1]
```diff
@@ -321,3 +323,17 @@ impl<T: Config> Xls20MintRequest for Pallet<T> {
...
+
+impl<T: Config> Xls20Ext for Pallet<T> {
+	fn get_xls20_token_id(token_id: TokenId) -> Option<seed_primitives::xrpl::Xls20TokenId> {
+		// TODO: update this
+		Xls20TokenMap::<T>::get(token_id.0, token_id.1).map(|id| (&id[..32]).try_into().unwrap())
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn set_xls20_token_id(token_id: TokenId, xls20_token_id: seed_primitives::xrpl::Xls20TokenId) {
+		let mut expanded_xls20_token_id = [0_u8; 64]; // TODO: update this
```
Comment (JasonTulp) [inline pallet/xls20/src/lib.rs:335]: TODO
Can just insert directly once migration PR is merged

File: pallet/xrpl-bridge/src/benchmarking.rs [hunk 2]
```diff
@@ -325,6 +326,65 @@ benchmarks! {
...
+	withdraw_nft {
+		let alice = account::<T>("Alice");
+		let door_address: XrplAccountId  = [1u8; 20].into();
+		let tx_fee = 100;
+		let nftoken_sell_offer = [2_u8; 32];
+		let destination = XrplAccountId::from_slice(b"6490B68F1116BFE87DDD");
+		// Note - had to do this since when CI running all tests, NextEventProofId has been incremented
+		// by one already, probably from a prior test execution. Keeping it simple as this to satisfy
+		// both benchmarks and tests since we just need this for verification purpose only.
+		let proof_id = if cfg!(test) { 1 } else { 0 };
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/benchmarking.rs:339]: This is to do with the impl of `XrplBridgeToEthyAdapter` for `MockEthyAdapter` in xrpl-bridge mock
Might be better to just change the mock impl to return `Ok(0)` and change the tests that assert the `WithdrawRequest` event

File: pallet/xrpl-bridge/src/lib.rs [hunk 3]
```diff
@@ -1383,6 +1410,35 @@ impl<T: Config> Pallet<T> {
...
+		who: T::AccountId,
+		collection_id: CollectionUuid,
+		serial_number: SerialNumber,
+		destination: XrplAccountId,
+	) -> DispatchResult {
+		let xls20_compatible = T::NFTExt::get_cross_chain_compatibility(collection_id)?;
+		ensure!(xls20_compatible.xrpl == true, Error::<T>::Xls20Incompatible);
+
+		// burn the nft
+		T::NFTExt::do_burn(who, collection_id, serial_number)?;
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/lib.rs:1424]: Might be better to transfer the NFT to the Xls20 palletId instead, maybe worth discussing. Not too sure which approach is best.

Comment (JasonTulp) [inline pallet/xrpl-bridge/src/lib.rs:1424]: Might be better to transfer the NFT to the Xls20 palletId instead, maybe worth discussing. Not too sure which approach is best. It looks like nft-peg burns

Comment (JasonTulp) [reply pallet/xrpl-bridge/src/lib.rs:1424]: It looks like nft-peg burns

File: pallet/xrpl-bridge/src/tests.rs [hunk 4]
```diff
@@ -4088,3 +4092,232 @@ fn generate_nft_accept_offer_failure_scenarios() {
...
+			OriginChain::Root,
+			CrossChainCompatibility { xrpl: true },
+		)
+		.unwrap();
+
+		assert_ok!(<Test as Config>::NFTExt::do_mint(alice(), nft_collection_id, 1, None,));
+
+		System::reset_events();
+		// withdraw by other than the owner
+		assert_err!(
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/tests.rs:4291]: nit, should use assert_noop for all error asserts

File: pallet/xrpl-bridge/src/tests.rs [hunk 5]
```diff
@@ -4088,3 +4092,232 @@ fn generate_nft_accept_offer_failure_scenarios() {
...
+			metadata_scheme,
+			None,
+			OriginChain::Root,
+			CrossChainCompatibility { xrpl: true },
+		)
+		.unwrap();
+
+		assert_ok!(<Test as Config>::NFTExt::do_mint(alice(), nft_collection_id, 1, None,));
+
+		System::reset_events();
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/tests.rs:4289]: Why reset events?

File: pallet/xrpl-bridge/src/tests.rs [hunk 6]
```diff
@@ -4088,3 +4092,232 @@ fn generate_nft_accept_offer_failure_scenarios() {
...
+			None,
+			OriginChain::Root,
+			CrossChainCompatibility { xrpl: false },
+		)
+		.unwrap();
+
+		assert_ok!(<Test as Config>::NFTExt::do_mint(alice(), nft_collection_id, 1, None,));
+
+		// withdraw nft
+		System::reset_events();
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/tests.rs:4225]: Don't need event reset

File: pallet/xrpl-bridge/src/weights.rs [hunk 7]
```diff
@@ -18,7 +18,7 @@
 //! Autogenerated weights for pallet_xrpl_bridge
 //!
 //! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 4.0.0-dev
-//! DATE: 2024-10-21, STEPS: `50`, REPEAT: 20, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! DATE: 2024-10-30, STEPS: `50`, REPEAT: 20, LOW RANGE: `[]`, HIGH RANGE: `[]`
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/weights.rs:21]: Need to run benchmarks. Might be easiest to run them for both pallets in the base PR once these are all merged

File: pallet/xrpl-bridge/src/tests.rs [hunk 8]
```diff
@@ -4088,3 +4092,232 @@ fn generate_nft_accept_offer_failure_scenarios() {
...
+			1_u32,
+			200_u32
+		));
+
+		// setup pallet-nft
+		let collection_name = BoundedVec::truncate_from("test".as_bytes().to_vec());
+		let collection_owner = alice();
+		let metadata_scheme = MetadataScheme::try_from(b"example.com/metadata".as_slice()).unwrap();
+
+		let nft_collection_id = <Test as Config>::NFTExt::do_create_collection(
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/tests.rs:4132]: You can use the NFTBuilder from pallets/nft/src/test_utils to easily create NFT collection

File: pallet/xrpl-bridge/src/benchmarking.rs (hunk 2 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/benchmarking.rs]: [dc2df18](https://github.com/futureversecom/trn-seed/pull/898/commits/dc2df180e6e293c133eadb9cf250939970229833)

File: pallet/xrpl-bridge/src/tests.rs (hunk 4 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/tests.rs]: [dc2df18](https://github.com/futureversecom/trn-seed/pull/898/commits/dc2df180e6e293c133eadb9cf250939970229833)

File: pallet/xrpl-bridge/src/tests.rs (hunk 5 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/tests.rs]: [dc2df18](https://github.com/futureversecom/trn-seed/pull/898/commits/dc2df180e6e293c133eadb9cf250939970229833)

File: pallet/xrpl-bridge/src/tests.rs (hunk 6 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/tests.rs]: [dc2df18](https://github.com/futureversecom/trn-seed/pull/898/commits/dc2df180e6e293c133eadb9cf250939970229833)

File: pallet/xrpl-bridge/src/tests.rs (hunk 8 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/tests.rs]: [c96f869](https://github.com/futureversecom/trn-seed/pull/898/commits/c96f8697558878b612ab727cf66c29632fd337e0)

File: pallet/xrpl-bridge/src/lib.rs (hunk 3 above)
Comment (JasonTulp) [reply pallet/xrpl-bridge/src/lib.rs:1424]: Went with Burn on the deposits to match

File: pallet/xrpl-bridge/src/lib.rs [hunk 9]
```diff
@@ -850,6 +863,20 @@ pub mod pallet {
...
 		}
+
+		/// Withdraw NFT to XRPL
+		#[pallet::call_index(17)]
+		#[pallet::weight(T::WeightInfo::withdraw_nft())]
+		#[transactional]
+		pub fn withdraw_nft(
+			origin: OriginFor<T>,
+			collection_id: CollectionUuid,
+			serial_number: SerialNumber,
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/lib.rs:874]: This should be token_id

File: pallet/common/src/lib.rs [hunk 10]
```diff
@@ -553,6 +572,18 @@ pub trait NFTExt {
 	fn get_collection_owner(
 		collection_id: CollectionUuid,
 	) -> Result<Self::AccountId, DispatchError>;
+
+	/// Burn a token
+	fn do_burn(
```
[the remark this answers is not in the document]
Comment (surangap) [reply pallet/common/src/lib.rs]: [e9de4ee](https://github.com/futureversecom/trn-seed/pull/898/commits/e9de4ee1002331ac66652cbd3f2aa81f6bfee2cd)

File: pallet/xls20/src/lib.rs (hunk 1 above)
Comment (surangap) [reply pallet/xls20/src/lib.rs]: [e9de4ee](https://github.com/futureversecom/trn-seed/pull/898/commits/e9de4ee1002331ac66652cbd3f2aa81f6bfee2cd)

File: pallet/xrpl-bridge/src/lib.rs (hunk 9 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/lib.rs]: [e9de4ee](https://github.com/futureversecom/trn-seed/pull/898/commits/e9de4ee1002331ac66652cbd3f2aa81f6bfee2cd)

File: pallet/xrpl-bridge/src/lib.rs [hunk 11]
```diff
@@ -74,6 +77,7 @@ type AccountOf<T> = <T as frame_system::Config>::AccountId;
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
+	use seed_primitives::TokenId;
```
Comment (JasonTulp) [inline pallet/xrpl-bridge/src/lib.rs:80]: nit, already imported above

**Review by JasonTulp — APPROVED** [review]
lgtm, nice work

File: pallet/xrpl-bridge/src/lib.rs (hunk 11 above)
Comment (surangap) [reply pallet/xrpl-bridge/src/lib.rs]: [96a6ccc](https://github.com/futureversecom/trn-seed/pull/898/commits/96a6ccc265caf391726aa01e8b284398567ba051)

**Pull request merged**